pubsub/google/cloud/pubsub_v1/publisher/client.py

```diff
@@ -167,44 +211,63 @@ def target(self):
+        """ Get an existing sequencer or create a new one given the (topic,
+            ordering_key) pair.
+        """
+        sequencer_key = (topic, ordering_key)
+        sequencer = self._sequencers.get(sequencer_key)
+        if sequencer is None:
+            if ordering_key == "":
+                # TODO: Consider implementing _delete_sequencer for unordered
+                # sequencer.
+                sequencer = unordered_sequencer.UnorderedSequencer(self, topic)
+            else:
+                sequencer = ordered_sequencer.OrderedSequencer(
+                    self, topic, ordering_key, self._delete_sequencer
+                )
+            self._sequencers[sequencer_key] = sequencer
+
+        return sequencer
 
-        This will create a new batch if ``create=True`` or if no batch
-        currently exists.
+    def resume_publish(self, topic, ordering_key):
+        """ Resume publish on an ordering key that has had unrecoverable errors.
 
         Args:
-            topic (str): A string representing the topic.
-            create (bool): Whether to create a new batch. Defaults to
-                :data:`False`. If :data:`True`, this will create a new batch
-                even if one already exists.
-            autocommit (bool): Whether to autocommit this batch. This is
-                primarily useful for debugging and testing, since it allows
-                the caller to avoid some side effects that batch creation
-                might have (e.g. spawning a worker to publish a batch).
+            topic (str): The topic to publish messages to.
+            ordering_key: A string that identifies related messages for which
+                publish order should be respected.
 
-        Returns:
-            ~.pubsub_v1._batch.Batch: The batch object.
+        Raises:
+            RuntimeError:
+                If called after publisher has been stopped by a `stop()` method
+                call.
+            ValueError:
+                If the topic/ordering key combination has not been seen before
+                by this client.
         """
-        # If there is no matching batch yet, then potentially create one
-        # and place it on the batches dictionary.
-        if not create:
-            batch = self._batches.get(topic)
-            if batch is None:
-                create = True
-
-        if create:
-            batch = self._batch_class(
-                autocommit=autocommit,
-                client=self,
-                settings=self.batch_settings,
-                topic=topic,
-            )
-            self._batches[topic] = batch
+        with self._batch_lock:
+            if self._is_stopped:
+                raise RuntimeError("Cannot resume publish on a stopped publisher.")
 
-        return batch
+            sequencer_key = (topic, ordering_key)
+            sequencer = self._sequencers.get(sequencer_key)
+            if sequencer is None:
```

Comment:
Is it necessary for this to be an error? Seems like it is reasonable just to silently let it through in this case. We want to be able to handle the case where one publishes messages for a key, there are errors, and then they never resume for the key or publish on it again.

Reply:
The scenario you describe is already allowed and isn't related to this error. If a batch on a key errors out, the ordered_sequencer for that topic/key is marked paused. Messages with other topic/key combos will get in just fine.

It's odd to resumePublish() for a topic/ordering_key pair that hasn't been seen before right? I think that should be an error.

Reply:
I think what I'm getting at is that we should delete the sequencer for that ordering key. Otherwise, if one never calls resume, then we will keep the sequencer around indefinitely, essentially leaking the memory.

Reply:
Well, I'm using an empty sequencer as a sentinel to mark that the ordering key has been paused. If we deleted the sequencer, we wouldn't be able to distinguish between a paused key and a key that's never been seen before. Yes, the amount of data in the publisher client grows, but it's necessary to implement pause/resume, so I would not call it a leak.

Reply:
I guess it's an artifact of how this is implemented in that case. I still don't think it needs to be an error. I think of it as a no-op. This is the approach we've taken in other languages and so I think we should be consistent here.

Reply:
Okay, made the changes:
1) Do NOT throw if an unknown topic-key pair is passed in. Instead log it.
2) Do throw if resume_publish is called without message ordering enabled.